lib/fxpay/index.js

```diff
@@ -34,63 +34,72 @@
-
-      receipts.all(function(error, allReceipts) {
+    return new Promise(function(resolve, reject) {
+      utils.getAppSelf(function(error, appSelf) {
         if (error) {
-          return callback(error, defaultProductInfo);
+          return reject({error: error, productInfo: defaultProductInfo});
         }
+        if (!appSelf) {
+          return reject({error: errors.PayPlatformUnavailable(
+                              'mozApps.getSelf() required for receipts'),
+                         productInfo: defaultProductInfo});
+        }
+        var allAppReceipts = [];
 
-        allReceipts.forEach(function(receipt) {
-          var storedata = receipts.checkStoreData(receipt);
-          if (!storedata) {
-            settings.log.info(
-              'ignoring receipt with missing or unparsable storedata');
-            return;
-          }
-          if (storedata.inapp_id) {
-            settings.log.info('ignoring in-app receipt with storedata',
-                              storedata);
-            return;
+        receipts.all(function(error, allReceipts) {
+          if (error) {
+            return reject({error: error, productInfo: defaultProductInfo});
           }
-          allAppReceipts.push(receipt);
-        });
 
-        settings.log.info('app receipts found:', allAppReceipts.length);
-
-        var appReceipt;
-
-        if (allAppReceipts.length === 0) {
-          return callback(errors.AppReceiptMissing(
-                              'no receipt found in getSelf()'),
-                              defaultProductInfo);
-        } else if (allAppReceipts.length === 1) {
-          appReceipt = allAppReceipts[0];
-          settings.log.info('Installed receipt:', appReceipt);
-          return receipts.validateAppReceipt(appReceipt,
-                                             function(error, productInfo) {
-            settings.log.info('got verification result for', productInfo);
-            callback(error, productInfo);
+          allReceipts.forEach(function(receipt) {
+            var storedata = receipts.checkStoreData(receipt);
+            if (!storedata) {
+              settings.log.info(
+                'ignoring receipt with missing or unparsable storedata');
+              return;
+            }
+            if (storedata.inapp_id) {
+              settings.log.info('ignoring in-app receipt with storedata',
+                                storedata);
+              return;
+            }
+            allAppReceipts.push(receipt);
           });
-        } else {
-          // TODO: support multiple app stores? bug 1134739.
-          // This is an unlikely case where multiple app receipts are installed.
-          return callback(errors.NotImplementedError(
+
+          settings.log.info('app receipts found:', allAppReceipts.length);
+
+          var appReceipt;
+
+          if (allAppReceipts.length === 0) {
+            return reject({error: errors.AppReceiptMissing(
```

Comment:
Maybe this would be more readable:

```JS
return reject({
  error: errors.AppReceiptMissing('no receipt found in getSelf()'),
  productInfo: defaultProductInfo,
});
```